### PR TITLE
fix(codedeploy): allow the install agent script's commands to exit with errors

### LINKED ISSUE #12764: (codedeploy): addExecuteFileCommand with installAgent set to true fails
When using `addExecuteFileCommand` for EC2 user data on an ASG associated with a Code Deploy deployment group with `installAgent` set to true, the generated user data script will fail if the dependencies are already installed/install commands exit with non zero statuses.

### Reproduction Steps
```ts
const asg = new AutoScalingGroup(this, "ASG", { ...props });

asg.userData.addExecuteFileCommand({ filePath: "some/path" });

new ServerDeploymentGroup(this, "DeploymentGroup", {
    autoScalingGroups: [asg],
    installAgent: true,
});
```
### What did you expect to happen?

- My file to be executed and the Code Deploy agent + dependencies to be installed.

### What actually happened?

- My file executed, but the script failed on installing the Code Deploy agent + dependencies 

### Other
The generated user data script looks like this:

```sh
#!/bin/bash
set -e
chmod +x 'some/path'
'some/path' undefined
PKG_CMD=`which yum 2>/dev/null`
if [ -z "$PKG_CMD" ]; then
PKG_CMD=apt-get
else
PKG=CMD=yum
fi
$PKG_CMD update -y
$PKG_CMD install -y ruby2.0
if [ $? -ne 0 ]; then #won't be run - script will fail out
$PKG_CMD install -y ruby
fi
$PKG_CMD install -y awscli
TMP_DIR=`mktemp -d`
cd $TMP_DIR
aws s3 cp s3://aws-codedeploy-eu-west-1/latest/install . --region eu-west-1
chmod +x ./install
./install auto
rm -fr $TMP_DIR
```

The [install agent code](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-codedeploy/lib/server/deployment-group.ts#L350-L368) relies on exit codes of commands to determine what to do, whilst the [`addExecuteFileCommand` code](https://github.com/aws/aws-cdk/blob/a72cfbd5d0f9af87aacf0657a6fc370ce2a23c55/packages/%40aws-cdk/aws-ec2/lib/user-data.ts#L165-L170) adds a `set -e` to fail the script if the file fails to run, but never swaps the flag back.

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by skinny85**
The script that installed the CodeDeploy agent on the hosts in the ServerDeploymentGroup relies on some commands failing,
and checking their return codes.
Unfortunately, calling the `addExecuteFileCommand()` method on the ASG sets the 'set -e' flag,
which makes the script fail.

Explicitly call the 'set +e' command in the script as the first instruction.

Fixes #12764

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts [hunk 1]
```diff
@@ -348,6 +348,7 @@ export class ServerDeploymentGroup extends ServerDeploymentGroupBase {
     switch (asg.osType) {
       case ec2.OperatingSystemType.LINUX:
         asg.addUserData(
+          'set +e', // make sure we don't exit on command errors - we check the return codes below
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts:351]: Feels like this should be in a subshell, or should be reversed afterwards (but a subshell feels safer). It's risky to switch off a global setting without switching it back.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts]: @rix0rrr I didn't like the way the sub-shell solution worked, so I did something different - I minimized the places where we use `set +e`, and made the script work with `set -e` (which I think is the correct thing to do).

Let me know if this works!

File: packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts [hunk 2]
```diff
@@ -348,15 +348,20 @@ export class ServerDeploymentGroup extends ServerDeploymentGroupBase {
...
+          'set +e', // make sure we don't exit on the `which` failing
           'PKG_CMD=`which yum 2>/dev/null`',
+          'set -e', // continue with failing on error
           'if [ -z "$PKG_CMD" ]; then',
           'PKG_CMD=apt-get',
           'else',
           'PKG=CMD=yum',
           'fi',
           '$PKG_CMD update -y',
+          'set +x', // make sure we don't exit on the next command failing (we check its exit code below)
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts:360]: Did you mean `set +e` ?

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codedeploy/lib/server/deployment-group.ts]: Yep. Thanks, great catch!

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**